### ISSUE TITLE
Question: How to calculate total & elapsed RenderEncode time for vkQueueSubmit()

### ISSUE THREAD

**Issue by SRSaunders** [KhronosGroup/MoltenVK#1979]
When running RBDoom3BFG on Windows and Linux using the [nvrhi](https://github.com/NVIDIAGameWorks/nvrhi) Vulkan backend, I can achieve a max of about ~90% GPU utilization (using my rx6600xt gpu) under heavy load.  On macOS using MoltenVK, I can approach a max of about 70% utilization with the same gpu, dropping to about 40% under heavy load.  This seems counter-intuitive, but may be explained by the need for MoltenVK to perform on-the-fly Vulkan to Metal command buffer encoding, which is likely eating up a significant portion of the frame time.  To help me understand what is going on, I would like to instrument my code to show total per-frame command buffer encoding time for a given `vkQueueSubmit()` call.  Within each submit there can be many Vulkan renderpasses (~2400 per frame), so there will likely be many Metal encoding steps per frame.

Note that I am using `synchronousQueueSubmits = false`, with triple buffering and external synchronization.

I have looked at using `vkGetPerformanceStatisticsMVK()` and the `MVKQueuePerformance` struct, but I don't see an entry for RenderEncoding.  I only see `mtlCommandBufferCompletion`, which may not be what I need.

Is there some way I can either:
1. Register and receive a callback when all RenderEncoding steps are complete for a given `vkQueueSubmit()`?, or
2. Access a per-frame performance timer for this (e.g. total and/or elapsed time)?

As an aside, I am curious about the RenderEncode sequencing in MoltenVK.  For a given Vulkan command buffer submission, are the Metal encode steps handled sequentially, or does MoltenVK attempt to parallelize this in any way?  Also, are the Metal encode operations 1:1 with Vulkan renderpasses?

**Comment by billhollings** [comment]
There are a few options for tracing time and load in MoltenVK.

1. You can make use of Apple's [HUD display](https://developer.apple.com/documentation/xcode/monitoring-your-metal-apps-graphics-performance/), which provides a variety of loading and timing stats per frame. There are several ways to enable the HUD, including environment variables.

2. You can use the Xcode Instruments tool to attach to your running app and dig out slow areas of execution in MoltenVK in your app.

3. If you want something specific, such as time to encode a Vulkan Command Buffer before committing it, you can modify the appropriate area in MoltenVK code, and use

``` 
uint64_t startTime = mvkGetTimestamp();
...
uint64_t elapsedTime = mvkGetElapsedNanoseconds(startTime);
```

to measure the elapsed time between two events, and then log it using `MVKLogInfo()` or `printf()`. Search the MoltenVK codebase for `mvkGetElapsedNanoseconds()` to see examples of it in use.

**Comment by SRSaunders** [comment]
Thanks for your response, but I have been using your suggestions 2 and 3 (to some extent) already.  While I have not used Apple's HUD before, the game already has a built-in HUD based on [imgui:](https://github.com/ocornut/imgui)

![Screenshot 2023-07-14 at 3 01 57 PM](https://github.com/KhronosGroup/MoltenVK/assets/82544213/7ae7056f-f042-4553-934e-71112146906b)


I was hoping there was a built-in mechanism to extract RenderEncode times from MoltenVK, so I could use it in release builds with the game's built-in HUD, and optionally with the game's built-in profiler based on [Optick](https://github.com/bombomby/optick).  For now I will pursue your option 3 for my dev-only builds.  As you can see from the above screen cap, the stats indicate the game should be running faster (greater fps) with lower CPU sync and GPU idle times, i.e. the game stats falsely imply the game is waiting on both CPU and GPU at the same time.  The missing information is likely MoltenVK's RenderEncode time on the CPU (as can be observed with Xcode Instruments).

Would you be interested in adding a struct member to `MVKQueuePerformance` that captures RenderEncode times in official MoltenVK builds?

**Comment by billhollings** [comment]
> Would you be interested in adding a struct member to MVKQueuePerformance that captures RenderEncode times in official MoltenVK builds?

Yes. We can add that fairly simply I should think.

**Comment by SRSaunders** [comment]
Good question.  On the surface I think option 2 above is my preference, but I would like to understand how multiple command buffer encode/submit operations from a single `vkQueueSubmit()` interact/overlap with the startup of GPU rendering.  If MoltenVK behaves close to a native implementation in this regard, then I think we will be okay - i.e. are they submitted as available or batched up and submitted together?

The game HUD you see displayed above uses elapsed time for CPU measurements in the left column, and Vulkan timer queries for GPU rendering measurements in the right column.  The one exception is that GPU idle time is calculated by subtracting GPU Total time (a Vulkan timer query) from the total frame period on the CPU side.  There is no implicit synchronization between LHS and RHS timers, other than the CPU Sync time on the left hand side.  This measures the CPU frame wait time on a Vulkan timeline semaphore for a barrier command buffer completion on the GPU (this is how nvrhi handles Vulkan sync).

I would like to take MoltenVK's total Vulkan-to-Metal render encoding time from the previous `vkQueueSubmit()` and subtract it from CPU Sync to give a more accurate assessment of where CPU time is being spent on a per-frame basis.  Ideally, after subtracting the total render encoding time, it should make the CPU Sync time approximately the same as one would observe for the same conditions on Windows or Linux.  Another way to think about it is adjusting MoltenVK's `vkQueueSubmit()` to be semantically and timing-wise equivalent to a native Vulkan implementation.

Does this make sense?  I realize that _approximate_ is an important term here since Vulkan and Metal concepts are not exactly aligned.  But best efforts are fine for this level of timing information.

**Comment by billhollings** [comment]
> are they submitted as available or batched up and submitted together

By default, MoltenVK encodes all the Vulkan command buffers in a  into a single `MTLCommandBuffer` for each `vkQueueSubmit()`, because it was found that this had the best all-around performance. However, this means the GPU does not start executing any of the encoded commands until all the Vulkan command buffers in the `vkQueueSubmit()` have been encoded into Metal. The results of this will depend on how your app is performance bound in time, and how many Vulkan command buffers you submit in each `vkQueueSubmit()`.

Have you tried playing with the `prefillMetalCommandBuffers` (env var `MVK_CONFIG_PREFILL_METAL_COMMAND_BUFFERS`) options? You can use this to create a 1:1 relationship between Vulkan and Metal command buffers, cause each of these `MTLCommandBuffer` to be encoded and submitted earlier, and therefore have those commands hit the GPU earlier. The downside has been potentially more churn within the Metal driver, causing slowdowns overall, but the modern Metal drivers may be getting better at this churn.

I'd be very interested to hear the results of your testing of these options in your scenario.

Finally, if you start to play with this, pay attention to the related `maxActiveMetalCommandBuffersPerQueue` (env var `MVK_CONFIG_MAX_ACTIVE_METAL_COMMAND_BUFFERS_PER_QUEUE`)) setting, to avoid blocking on `MTLCommandBuffer` allocation from pool, if you require a lot (more than 64) concurrently allocated`MTLCommandBuffers` per queue.

**Comment by SRSaunders** [comment]
Thanks very much for your explanation re batch vs sequential submission - very useful to know.  I had already tried the `prefillMetalCommandBuffers` options but as you indicate above the default setting was the best.  My situation is much simpler in that RBDoom3BFG currently uses a single Vulkan command buffer containing many renderpasses (~2300).  As would be expected the prefill options will likely not help in this case.

I am still looking for a way to optimize MoltenVK's command buffer render encode time.  Up to now I have implemented triple buffering for max overlap/throughput combined with MoltenVK's async queue submission option.  Here are a few screenshots of a worst case scenario in the game:

Screenshot 1: Zoom in on the HUD in right top corner.  I have prototyped the `mtlCommandBufferEncode` perf counter as discussed above (see "Encode" line in the HUD).  The Game falls below 60 fps with CPU loading of 100% primarily caused by MoltenVK's **encode time of 24.4 msec**.  Note the total **GPU rendering time of 13.4 msec** (GPU only 41% loaded at 32 fps) which is less than the 16 msec needed for 60 fps, i.e. when using MoltenVK the game is CPU bound not GPU bound.

![Screenshot 2023-07-17 at 6 37 40 PM](https://github.com/KhronosGroup/MoltenVK/assets/82544213/d2a40533-2214-48dd-b876-1ab874ba3de5)

Screenshot 2: An Xcode Instruments trace of the game running with triple buffering, and `synchronousQueueSubmits = false`.  Notice that `vkQueueSubmit()` is called in parallel to GPU rendering, but doesn't start encoding until the driver and GPU are finished processing the previous frame (see **_dispatch_workloop_worker_thread**).  Why can't encoding start earlier when `vkQueueSubmit()` is first called?  Is there a threading limitation which prevents parallel encoding?

![Screenshot 2023-07-17 at 10 03 54 PM](https://github.com/KhronosGroup/MoltenVK/assets/82544213/67bac070-df99-407d-bf1f-39abaf2935d5)

Screenshot 3: An expanded Xcode Instruments trace showing the impact of thousands of renderpasses in a single command buffer.  There appears to be a significant loss of efficiency with Xcode Instruments showing 40-50% idle time between the many RenderEncoder steps.  Is this an inherent issue with using many renderpasses with MoltenVK?

![Screenshot 2023-07-17 at 10 09 19 PM](https://github.com/KhronosGroup/MoltenVK/assets/82544213/3904a3f3-3b2e-44fe-9991-25aca558a737)

**Comment by billhollings** [comment]
> I had already tried the `prefillMetalCommandBuffers` options

Did you try `MVK_CONFIG_PREFILL_METAL_COMMAND_BUFFERS_STYLE_IMMEDIATE_ENCODING` or `MVK_CONFIG_PREFILL_METAL_COMMAND_BUFFERS_STYLE_IMMEDIATE_ENCODING_NO_AUTORELEASE`? This will encode the command onto the `MTLCommandBuffer` immediately when it is added, avoiding the delay until all the commands have been added to the Vulkan command buffer.

> Why can't encoding start earlier when vkQueueSubmit() is first called?

This is indeed odd. The connection between say, CPU-orange/GPU-purple and the subsequent CPU-green, seems to indicate the blocking of a CPU based `VkSemaphore` (`MVKSemaphoreEmulated` or `MVKTimelineSemaphoreEmulated`) or possibly 

What appears in the log at "Vulkan semaphores using..."?

And in the trace snapshot, the top bar seems to indicate the lifetime of the `MTLCommandBuffer`. What does the bottom bar represent? You seem to indicate it is the encoding? Or is it the activities of the commit?

![Screenshot 2023-07-18 at 6 58 46 PM](https://github.com/KhronosGroup/MoltenVK/assets/356402/1c244af3-49b1-4810-b64d-63d3ae216327)

> Screenshot 3...showing 40-50% idle time between the many RenderEncoder steps

Similar question, does the bottom bar represent the activities of encoding onto the `MTLCommandBuffer`?

**Comment by SRSaunders** [comment]
> Did you try `MVK_CONFIG_PREFILL_METAL_COMMAND_BUFFERS_STYLE_IMMEDIATE_ENCODING` or `MVK_CONFIG_PREFILL_METAL_COMMAND_BUFFERS_STYLE_IMMEDIATE_ENCODING_NO_AUTORELEASE`?

`MVK_CONFIG_PREFILL_METAL_COMMAND_BUFFERS_STYLE_NO_PREFILL` = 32 fps
`MVK_CONFIG_PREFILL_METAL_COMMAND_BUFFERS_STYLE_IMMEDIATE_ENCODING` = 26 fps
`MVK_CONFIG_PREFILL_METAL_COMMAND_BUFFERS_STYLE_IMMEDIATE_ENCODING_NO_AUTORELEASE` = 15 fps

> What appears in the log at "Vulkan semaphores using..."?

[mvk-info] Vulkan semaphores using MTLEvent.

> And in the trace snapshot, the top bar seems to indicate the lifetime of the MTLCommandBuffer. What does the bottom bar represent? You seem to indicate it is the encoding? Or is it the activities of the commit?

In screenshot 2, I believe the top bar is elapsed time and the bottom bar shows the compute activities of the dispatched thread responsible for asynchronous encoding (note `synchronousQueueSubmits = false`).  This is default presentation of Xcode's Game Performance Instrument.  I have not customized anything in that view.

> Similar question, does the bottom bar represent the activities of encoding onto the MTLCommandBuffer?

Screenshot 3 is simply a magnified version of screenshot 2, where you can see the details of the bottom bar.   I was hoping you would recognize the labeling, since it comes directly from MoltenVK, e.g. **vkCmdBeginRenderPass RenderEncoder** comes from **MVKCommandBuffer.mm** line 1184:

```
NSString* mvkMTLRenderCommandEncoderLabel(MVKCommandUse cmdUse) {
    switch (cmdUse) {
	case kMVKCommandUseBeginRendering:                  return @"vkCmdBeginRendering RenderEncoder";
-->     case kMVKCommandUseBeginRenderPass:                 return @"vkCmdBeginRenderPass RenderEncoder";
        case kMVKCommandUseNextSubpass:                     return @"vkCmdNextSubpass RenderEncoder";
```

**Comment by billhollings** [comment]
>`MVK_CONFIG_PREFILL_METAL_COMMAND_BUFFERS_STYLE_NO_PREFILL = 32 fps`
>`MVK_CONFIG_PREFILL_METAL_COMMAND_BUFFERS_STYLE_IMMEDIATE_ENCODING = 26 fps`
>`MVK_CONFIG_PREFILL_METAL_COMMAND_BUFFERS_STYLE_IMMEDIATE_ENCODING_NO_AUTORELEASE = 15 fps`

Thanks. Interesting and odd. These are counterintuitive results for a single command buffer. Immediate encoding should be getting the commands into the `MTLCommandBuffer` sooner. I'll have to look into that at some point.

But nevertheless, this is probably not an issue for this situation.

> the bottom bar shows the activities during asynchronous encoding (note `synchronousQueueSubmits = false`).

There isn't really anything in MoltenVK that should be blocking the encoding. Almost no activity occurs between the ending of one renderpass and the beginning of the next. Beginning a renderpass has some setup activity, but nothing that should be taking almost as much time as encoding all the activity onto the renderpass takes. And the delay before command buffer encoding seem to be blocked until the framebuffer present activity finishes (which is why I was asking about `VkSemaphore`).

What do the Instruments traces in Screenshot 2 & 3 look like when `synchronousQueueSubmits = true`? I know setting this will add processing to you app rendering thread, and I'm not looking at it as a solution, but I'm curious about whether the dispatch thread is being periodically interrupted by the system to prioritize activities on other dispatch threads. We might not have the priorities set optimally for that dispatch thread, or perhaps we may have to move away from using GCD for this thread.

**Comment by SRSaunders** [comment]
Here are a few screen caps with `synchronousQueueSubmits = true`:

Screenshot 1: See the HUD - the encoding has been moved to the render thread (see "RB:" line) but with lower fps.  Note that async "Encode" = 0
![Screenshot 2023-07-19 at 12 04 07 PM](https://github.com/KhronosGroup/MoltenVK/assets/82544213/6343c59f-d3da-4558-baa2-d32781610c3f)

Screenshot 2: Queue submission is synchronous showing no overlap with driver/GPU.  Also there is no initial waiting for the start of encoding.  However, note the **elapsed encode duration is 46.7 msec**, with a **total encoding activity time of 26.34 msec**, i.e. the encoding inefficiency is still present.

Regarding the initial wait time when `synchronousQueueSubmits = false`, the only semaphores being used are: a) a queue signal semaphore at the end of the command buffer indicating to `vkQueuePresent()` that the frame is ready for presentation, and b) a timeline semaphore inserted by nvrhi indicating the end of the command buffer processing for CPU/GPU frame synchronization.  Could these be somehow influencing MoltenVK's encoding thread?

![Screenshot 2023-07-19 at 12 21 21 PM](https://github.com/KhronosGroup/MoltenVK/assets/82544213/cb902883-9052-4beb-aa49-9a851da7a68d)

Screenshot 3: Magnified version of screenshot 2, showing gaps between render encode activities.
![Screenshot 2023-07-19 at 12 23 48 PM](https://github.com/KhronosGroup/MoltenVK/assets/82544213/f3d91fe9-23db-4847-acce-37eceea47172)

I will be away for a few days, so I will not be able to respond until the end of the week.  I hope this is enough information for you to investigate what may be going on.

**Comment by SRSaunders** [comment]
Here's another screenshot regarding the initial wait when `synchronousQueueSubmits = false`.  It looks like the Metal driver is executing some kind of resource lock that corresponds to the waiting interval before encoding gets started for the next frame.  Is there a shared resource in MoltenVK (e.g. a generated Metal command buffer, a mapped memory buffer, etc) that is passed to the Metal driver but does not get released until driver processing is finished for that frame?  Could such a shared resource be blocking the startup of overlapped frame encoding?  Is MoltenVK designed to handle this kind of overlap, or is this a known limitation?  I'm trying to figure out if the game is doing something wrong regarding sync, or if MoltenVK has an issue with this use case.

![Screenshot 2023-07-21 at 9 37 56 PM](https://github.com/KhronosGroup/MoltenVK/assets/82544213/38b81b86-755d-43a7-bb2d-599a4658333f)

**Comment by SRSaunders** [comment]
Well, I think I have found the issue that delays the start of async encoding.  However, I still don't know why the encoding elapsed time is almost twice the encoding activity time.  But first steps first...

When `MVKQueueCommandBufferSubmission::execute()` calls `submitCommandBuffers()`, the following block of code is executed:

```
template <size_t N>
void MVKQueueFullCommandBufferSubmission<N>::submitCommandBuffers() {
	_queue->getPhysicalDevice()->startTimestampCorrelation(_cpuStart, _gpuStart);
	for (auto& cb : _cmdBuffers) { cb->submit(this, &_encodingContext); }
}
```

The call to `startTimestampCorrelation()` appears to be synchronous to the next Drawable Present, at least for my GPU (AMD Radeon 6600 on Intel running Ventura 13.5).  If I move the call to after the encoding step like this...

```
template <size_t N>
void MVKQueueFullCommandBufferSubmission<N>::submitCommandBuffers() {
	for (auto& cb : _cmdBuffers) { cb->submit(this, &_encodingContext); }
	_queue->getPhysicalDevice()->startTimestampCorrelation(_cpuStart, _gpuStart);
}
```

The Instruments trace now looks like the following, which is starting to gain some parallel encoding efficiency. One problem down, and one to go - why does encoding efficiency still appear to be slow?

![Screenshot 2023-07-30 at 10 52 10 PM](https://github.com/KhronosGroup/MoltenVK/assets/82544213/79928621-66f3-4afe-9415-0055bd725b43)

**Comment by billhollings** [comment]
> I think I have found the issue that delays the start of async encoding

Ah! Very interesting! And great research and detective work!

That `startTimestampCorrelation()` operation is not used on Apple Silicon, and, in the Vulkan context, it is only used for updating `VkPhysicalDeviceLimits::timestampPeriod`, which is certainly not expected to be queried anywhere that frequently, if at all. 

So, although Apple recommends using it on `MTLCommandBuffers`, we should move it. In fact, perhaps we should move it to `vkGetPhysicalDeviceProperties()` itself. I'll add a separate PR for that, and get it done soon.

**Comment by billhollings** [comment]
> I've add a separate PR #1983 for that, and I'll get it done soon, within the next week or two.

PR #1992 fixes issue #1983.

**Comment by SRSaunders** [comment]
> PR https://github.com/KhronosGroup/MoltenVK/pull/1992 fixes issue https://github.com/KhronosGroup/MoltenVK/issues/1983.

Thanks very much for fixing this.  I tried PR #1992 and it seems to work fine.

Would you also consider putting in the following change (as well as related definition/inits in **mvc_private_api.h** and **MVKDevice.mm** ) for a new `mtlCommandBufferEncode` queue performance counter, which was the original request for this issue?  Once you subtract `nextCAMetalDrawable` from this value, you get a pretty accurate view of encoding time, even when vsync is turned on (i.e. when `VK_PRESENT_MODE_FIFO_KHR` is enabled).

```
void MVKQueueCommandBufferSubmission::execute() {

	_queue->_submissionCaptureScope->beginScope();
	// If using encoded semaphore waiting, do so now.
	for (auto& ws : _waitSemaphores) { ws.first->encodeWait(getActiveMTLCommandBuffer(), ws.second); }

-->	// Capture start time before command encoding
-->	MVKDevice* mvkDev = _queue->getDevice();
-->	uint64_t startTime = mvkDev->getPerformanceTimestamp();

	// Submit each command buffer.
	submitCommandBuffers();

-->	// Capture end time for command buffer encode performance counter 
-->	mvkDev->addActivityPerformance(mvkDev->_performanceStatistics.queue.mtlCommandBufferEncode, startTime);

	// If using encoded semaphore signaling, do so now.
	for (auto& ss : _signalSemaphores) { ss.first->encodeSignal(getActiveMTLCommandBuffer(), ss.second); }

	// Commit the last MTLCommandBuffer.
	// Nothing after this because callback might destroy this instance before this function ends.
	commitActiveMTLCommandBuffer(true);
}
```

The deeper problem of why there is so much overhead during encoding still eludes me.  I have looked at the encoder thread execution time and it is not blocked except for very short periods, yet encoding elapsed time remains significantly greater than total encoding activity time.  From my investigations, it appears that `beginMetalRenderPass()` and/or `endMetalEncoding()` have overheads that are magnified when encoding thousands of small Vulkan RenderPasses in a single `vkQueueSubmit()`.  I am not sure why this is the case, or if any settings regarding memory allocation and/or deallocation or other things could be changed to reduce that overhead.  I am at the end of what I understand on this issue and ask for your help in answering this question.

**Comment by billhollings** [comment]
> I tried PR https://github.com/KhronosGroup/MoltenVK/pull/1992 and it seems to work fine.

Wonderful! Thanks for the feedback!

> Would you also consider putting in the following change 

I hear you. 

So...I have a set of modifications to the performance tracking environment, in flight, as part of a larger sync issue I'm trying to fix. It's a WIP and is intrusive, so I don't want to release it to MoltenVK until after the SDK release this week.

`MVKQueuePerformance` will now include:

```
typedef struct {
	MVKPerformanceTracker retrieveMTLCommandBuffer;     /** Retrieve a MTLCommandBuffer from a MTLQueue. */
	MVKPerformanceTracker commandBufferEncoding;        /** Encode a single VkCommandBuffer to a MTLCommandBuffer (excludes when configured for immediate prefilling). */
	MVKPerformanceTracker mtlCommandBufferExecution;    /** Execute a MTLCommandBuffer on the GPU, from commit to completion callback. */
	MVKPerformanceTracker retrieveCAMetalDrawable;      /** Retrieve next CAMetalDrawable from a CAMetalLayer. */
	MVKPerformanceTracker presentSwapchains;            /** Present the swapchains in a vkQueuePresentKHR() on the GPU, from commit to presentation callback. */
	MVKPerformanceTracker frameInterval;                /** Frame presentation interval (1000/FPS). */
} MVKQueuePerformance;
```

Note `commandBufferEncoding`. This is similar to what you are asking for, but is more granular in that it tracks stats for each `VkCommandBuffer -> MTLCommandBuffer` encoding (recall that many `VkCommandBuffers` can be encoded onto a single `MTLCommandBuffer`).

Given that, do you see additional value in also tracking performance on the total time taken to encode all the `VkCommandBuffers` in a `vkQueueSubmit()`, as  you request above? That would generally track the total time taken to encode each`MTLCommandBuffer`, which I suppose would form a nice complement to `mtlCommandBufferExecution`. 

If you think that would provide additional helpful performance info, I can easily add it as an additional member in `MVKQueuePerformance`, as you suggest above.

And then look for this functionality to be released to MoltenVK sometime the week after the SDK release.

**Comment by SRSaunders** [comment]
> This is similar to what you are asking for, but is more granular in that it tracks stats for each VkCommandBuffer -> MTLCommandBuffer encoding

If there are multiple VkCommandBuffers involved in a single `vkQueueSubmit()`, how would one access the `commandBufferEncoding` perf counter for each one?  Wouldn't you need an array to capture that?  I thought that `MVKPerformanceTracker` was a scalar struct.  Currently I am calling `vkGetPerformanceStatisticsMVK()` once per frame to capture the performance data I need for the game HUD.  Would the return value for `commandBufferEncoding.latestDuration` be the last one processed, even if it was part of multi-command buffer submit?  I presume it would default to the only one processed in the case of a single command buffer submit (note in my case a single command buffer is submitted per frame).  I am just trying to figure out the semantics and query frequency/timing for what you are proposing.

> do you see additional value in also tracking performance on the total time taken to encode all the VkCommandBuffers in a vkQueueSubmit()

I was trying to keep the semantics simple and consistent in my proposal since the remaining counters (e.g. `mtlCommandBufferExecution`, `retrieveCAMetalDrawable`, `presentSwapchains`, and `frameInterval`) all seem to be defined at the per-frame level.  So yes, I think it would be useful to have a total encoding time counter that exists at the per-frame level for consistency.

**Comment by billhollings** [comment]
> If there are multiple VkCommandBuffers involved in a single vkQueueSubmit(), how would one access the commandBufferEncoding perf counter for each one?

Performance statistics are provided as avg, min, max, and count. None of them track individual values (except latest). So the intention would be that if you had multiple `VkCommandBuffers` in a submit, the counts would be different between encoding and execution of a command buffer. For instance, if you had an average of 10 `VkCommandBuffers` per `vkQueueSubmit()`, the count for `commandBufferEncoding` would be 10 times that of `mtlCommandBufferExecution` (and you'd manually multiply to compare).

But, as you indicate, if you want to easily match `vkQueueSubmits()` to `MTLCommandBuffer` executions, then a separate `submitCommandBuffers` tracker might make that easier. I've gone ahead and added that now in my WIP.

**Comment by SRSaunders** [comment]
> if you want to easily match vkQueueSubmits() to MTLCommandBuffer executions, then a separate submitCommandBuffers tracker might make that easier. I've gone ahead and added that now in my WIP.

Thanks for adding this in.  My use case (a game HUD) requires per-frame actuals (i.e. `.latestDuration`) vs avg, min, or max.  Having an aggregate tracker for total `vkQueueSubmit()` encoding time would be very useful, perhaps even necessary to support this.  I look forward to official support in MoltenVK.

Lastly, do you have any thoughts on my comments re encoding overhead?  Is this something you are interested in pursuing, or do you feel my use case of many render passes is an edge case?  I'm wondering whether I should hold out hope for some solution here, or instead try a different path (e.g. try to reduce the number of render passes/encoding overhead by leveraging push constants or some other game re-architecture approach).

**Comment by billhollings** [comment]
> do you have any thoughts on my comments re encoding overhead
>...
> it appears that `beginMetalRenderPass()` and/or `endMetalEncoding()` have overheads that are magnified when encoding thousands of small Vulkan RenderPasses from a single `vkQueueSubmit()`. 

Thanks for highlighting this. Yes, we definitely want to be fast in this situation. I'll try to set aside some time to dig into possible bottlenecks in `beginMetalRenderPass()` and `endMetalEncoding()`.

**Comment by SRSaunders** [comment]
> Yes, we definitely want to be fast in this situation. I'll try to set aside some time to dig into possible bottlenecks in beginMetalRenderPass() and endMetalEncoding().

Thanks for agreeing to look into this.  It looks like MoltenVK's implementation of Vulkan dynamic rendering ultimately relies on Metal render passes, so perhaps this may indeed be a broader issue.

**Comment by billhollings** [comment]
PR #2009 includes additional performance measuring around `MTLCommandBuffers` and queue submits, as discussed [above](#issuecomment-1676458108).

**Comment by SRSaunders** [comment]
Thanks for letting me know.  I will check it out after returning from the long weekend.

Regarding two related topics from above:

1) Performance - I have found that using push constants instead of constant buffers yields a massive encoding and runtime performance improvement.  In the case of RBDoom3BFG the encoding time drops from ~23 msec to ~4 msec, and runtime execution on my 6600XT drops from ~13 msec to  ~4 msec.  I am not sure if this is due to reducing the number of render passes per frame (~2400 down to ~50), or if this change is more directly related to potential overheads in MoltenVK when using volatile constant buffers.  Do you have any insight into this?

2) timestampPeriod issues - is there any way to disable the timestampPeriod low pass filter that was implemented right at the end of the release cycle?  This negatively affects my Optick profiler integration which depends on an accurate measurement of this value for GPU/CPU timeline calibration just before beginning a profiling trace.  Averaging this value ruins the timeline calibration and GPU traces are now misaligned with CPU traces.  Could this filter either be removed or possibly disabled via a config setting?  When I get back I will raise a new issue about this.

**Comment by SRSaunders** [comment]
> PR #2009 includes additional performance measuring around `MTLCommandBuffers` and queue submits

Your new code seems to work well.  I have tested `submitCommandBuffers.latest - retrieveCAMetalDrawable.latest` to determine the total encode time for the latest `vkQueueSubmit()`.  I am subtracting the `CAMetalDrawable` retrieval time to properly handle the `vsync = on` case.

**Comment by SRSaunders** [comment]
Thanks for all your work on the original issue re perf counters, and your perf improvement for async vkQueueSubmits().  I will close this now.

I still am not sure about possible renderpass and/or volatile constant buffer overheads when used thousands of times per frame.  However I do have a solution using push constants which bypasses both of these and provides an amazing performance boost.  If this is an area you want to look at in the future we could open a new issue to track separately.
